main/src/ca/uwaterloo/flix/language/phase/unification/Zhegalkin.scala

```diff
@@ -0,0 +1,171 @@
+
+    override def toString: String = {
+      if (s.isEmpty) "Ø"
+      else if (s.isUniverse) "T"
+      else s match {
+        case CofiniteIntSet.Set(xs) => s"{${xs.mkString(", ")}}"
+        case CofiniteIntSet.Compl(xs) => s"~{${xs.mkString(", ")}}"
+      }
+    }
+  }
+
+  private val ZEmpty: ZhegalkinConstant = ZhegalkinConstant(CofiniteIntSet.empty)
+  private val ZUniverse: ZhegalkinConstant = ZhegalkinConstant(CofiniteIntSet.universe)
+
+  /** Represents a Zhegalkin term: c n x1 n x2 n ... n xn */
+  case class ZhegalkinTerm(cst: ZhegalkinConstant, vars: SortedSet[ZhegalkinVar]) {
+    override def toString: String =
+      if (vars.isEmpty)
+        cst.toString
+      else
+        s"$cst n ${vars.mkString(" n ")}"
+  }
+
+  /** Represents a Zhegalkin expr: c ? t1 ? t2 ? ... ? tn */
+  case class ZhegalkinExpr(cst: ZhegalkinConstant, terms: List[ZhegalkinTerm]) {
+    override def toString: String =
+      if (terms.isEmpty)
+        cst.toString
+      else
+        s"$cst ? ${terms.map(t => s"($t)").mkString(" ? ")}"
+  }
+
+  /**
+    * A smart constructor to filter empty intersections.
+    */
+  private def mkZhegalkinExpr(cst: ZhegalkinConstant, terms: List[ZhegalkinTerm]): ZhegalkinExpr =
+    ZhegalkinExpr(cst, terms.filter(t => !t.cst.s.isEmpty))
+
+  /** Returns the xor of the two Zhegalkin constants */
+  private def mkXor(c1: ZhegalkinConstant, c2: ZhegalkinConstant): ZhegalkinConstant = {
+    // a ? b = (a ? b) - (a n b) = (a ? b) n ~(a n b)
+    c1.union(c2).inter(c1.inter(c2).compl)
+  }
+
+  /** Returns the xor of the two Zhegalkin expressions. */
+  private def mkXor(z1: ZhegalkinExpr, z2: ZhegalkinExpr): ZhegalkinExpr = (z1, z2) match {
+    case (ZhegalkinExpr(c1, ts1), ZhegalkinExpr(c2, ts2)) =>
+      val c = mkXor(c1, c2)
+      val ts = (ts1 ++ ts2).groupBy(identity).collect { case (k, v) if v.size % 2 != 0 => k }.toList
+      mkZhegalkinExpr(c, ts)
+  }
+
+  /** Returns the complement of the Zhegalkin expr. */
+  private def zmkCompl(a: ZhegalkinExpr): ZhegalkinExpr =
+    // ~a = 1 ? a
+    mkXor(ZhegalkinExpr(ZUniverse, Nil), a)
+
+  //
+  // (c1 ? t11 ? t12 ? ... ? t1n) n (c2 ? t21 ? t22 ? ... ? t2m)
+  //   =   (c1  n (c2 ? t21 ? t22 ? ... ? t2m)
+  //     ? (t11 n (c2 ? t21 ? t22 ? ... ? t2m)
+  //     ? (t12 n (c2 ? t21 ? t22 ? ... ? t2m)
+  //
+  private def mkInter(z1: ZhegalkinExpr, z2: ZhegalkinExpr): ZhegalkinExpr = z1 match {
+    case ZhegalkinExpr(c1, ts1) =>
+      val zero = mkInterConstantExpr(c1, z2)
+      ts1.foldLeft(zero) {
+        case (acc, z) => mkXor(acc, mkInterTermExpr(z, z2))
+      }
+  }
+
+  //
+  // c n (c2 n x1 n x2 n ... n xn) = (c n c2) n x1 n x2 n ... n xn)
+  //
+  private def mkInterConstantTerm(c: ZhegalkinConstant, t: ZhegalkinTerm): ZhegalkinTerm = t match {
+    case ZhegalkinTerm(c2, vars) =>
+      ZhegalkinTerm(c.inter(c2), vars)
+  }
+
+  //
+  // c n (c2 ? t21 ? t22 ? ... ? t2m) = (c n c2) ? t21 ? t22 ? ... ? t2m
```

Comment:
Is this a flaw in GitHub or?

Reply:
My GitHub phone app is not displaying Unicode, sadly

Reply:
Thats my point: I also don't see the unicode-- so is it there or not?